pkg/csi_driver/controller.go

```diff
@@ -101,15 +106,61 @@ func (s *controllerServer) CreateVolume(ctx context.Context, req *csi.CreateVolu
 			return nil, status.Error(codes.AlreadyExists, err.Error())
 		}
 	} else {
+
+		// Get the reserved-IPV4-cidr from the request. It was not added in the service instance as it is only consumed by the controller
+		reservedIPV4CIDR, reservedIPV4CIDRPresent := req.GetParameters()[paramReservedIPV4CIDR]
+		if !reservedIPV4CIDRPresent {
+			return nil, fmt.Errorf("The required parameter reserved-ipv4-cidr was not provided")
+		}
+
+		// Prevents lock acquisition in case the CIDR is invalid
+		err = s.config.ipAllocator.ValidateCIDR(reservedIPV4CIDR)
+		if err != nil {
+			return nil, err
+		}
+
+		reservedIPRange, err := s.getUnreservedIPBlock(ctx, newFiler, reservedIPV4CIDR)
+		if err != nil {
+			return nil, err
+		}
+
+		// Adding the reserved IP range to the instance object. It is freed up when the instance is created as we would then be getting this value in the list instances response for subsequent requests
+		newFiler.Network.ReservedIpRange = reservedIPRange
+
 		// Create the instance
 		filer, err = s.config.fileService.CreateInstance(ctx, newFiler)
 		if err != nil {
 			return nil, status.Error(codes.Internal, err.Error())
 		}
 	}
+
+	//Free up the reserved IPV4 Range from the ipAllocator as we would be recieving this value in the listInstances call as the createInstance operation was completed
+	s.config.ipAllocator.Unreserve(filer.Network.ReservedIpRange)
 	return &csi.CreateVolumeResponse{Volume: fileInstanceToCSIVolume(filer, modeInstance)}, nil
 }
 
+// getUnreservedIPBlock returns the available IP in the cidr. It is synchronized in order to avoid data race while getting unreserved IP range in a CIDR
+func (s *controllerServer) getUnreservedIPBlock(ctx context.Context, filer *file.ServiceInstance, cidr string) (string, error) {
+	instances, err := s.config.fileService.ListInstances(ctx, filer)
+	reservedIPRanges := make(map[string]bool)
+	for _, instance := range instances {
+		reservedIPRanges[instance.Network.ReservedIpRange] = true
+	}
+	s.config.mutex.Lock()
+	defer s.config.mutex.Unlock()
+	if err != nil {
+		return "", status.Error(codes.Aborted, err.Error())
+	}
+
+	// Pass the reservedIPRanges obtained from ListInstances call. We have the IPRanges reserved at the service end in the IPAllocator.reserved list. The GetUnreservedIPBlock would merge these two lists to form a reserved list
+	unreservedIPBlock, err := s.config.ipAllocator.GetUnreservedIPBlock(cidr, reservedIPRanges)
+	if err != nil {
+		return "", err
+	}
+	s.config.ipAllocator.Reserve(unreservedIPBlock)
```

Comment:
I wouldn't recommend doing a "mutating" action like `Reserve` inside of a function called `[get]UnreservedIPBlock`, if its mutating its more of an `AllocateUnreservedIPBlock`. Naming is important for future maintainers to easily follow logic and prevent bugs from people misusing your functions